src/hypothesis/extra/datetime.py

```diff
@@ -15,111 +15,412 @@
+                        low = max(low, getattr(min_dt, name))
+                    if constrained_high:
+                        high = min([high, getattr(max_dt, name)])
+                    if name == 'year':
+                        val = cu.centered_integer_range(data, low, high, 2000)
+                    else:
+                        val = cu.integer_range(data, low, high)
+                    result[name] = val
+                    constrained_low = constrained_low and val == low
+                    constrained_high = constrained_high and val == high
+                return dt.datetime(tzinfo=min_dt.tzinfo, **result)
+            except ValueError:
+                # We tried to draw a non-existent time, e.g. Feb 31
+                continue
+
+    def do_draw(self, data):
+        if self.min_dt.tzinfo == self.max_dt.tzinfo:
+            # No need to pick a timezone; they're the same (or both naive)
+            min_dt = self.min_dt
+            max_dt = self.max_dt
+        else:
+            # Choose a timezone to work in on this draw
+            if cu.boolean(data):
+                min_dt = self.min_dt
+                try:
+                    max_dt = self.max_dt.astimezone(self.min_dt.tzinfo)
+                except OverflowError:  # pragma: no branch  # pytest internals
+                    max_dt = self.min_dt.tzinfo.localize(dt.datetime.max)
+            else:
+                max_dt = self.max_dt
+                try:
+                    min_dt = self.min_dt.astimezone(self.max_dt.tzinfo)
+                except OverflowError:
+                    min_dt = self.max_dt.tzinfo.localize(dt.datetime.min)
+
+        result = self._draw_naive(data, min_dt, max_dt)
+        if (not self.timezones) or (self.allow_naive and cu.boolean(data)):
+            assert self.min_dt.tzinfo is None
+            return result
+        timezones = list(self.timezones)
+        for _ in range(len(self.timezones)):
+            try:
+                to_tz = cu.choice(data, timezones)
+                if self.min_dt.tzinfo is None:
+                    return to_tz.normalize(to_tz.localize(result))
+                else:
+                    return to_tz.normalize(result.astimezone(to_tz))
+            except OverflowError:
+                timezones.remove(to_tz)
+        raise InvalidState('Reached impossible state')  # pragma: no cover
 
 
 @defines_strategy
-def datetimes(allow_naive=None, timezones=None, min_year=None, max_year=None):
+def datetimes(allow_naive=None, timezones=None, min_year=None, max_year=None,
+              min_datetime=None, max_datetime=None):
     """Return a strategy for generating datetimes.
 
     allow_naive=True will cause the values to sometimes be naive.
     timezones is the set of permissible timezones. If set to an empty
-    collection all timezones must be naive. If set to None all available
+    collection all timezones will be naive. If set to None all available
     timezones will be used.
```

Comment:
It is a bit confusing!  I don't think there's a simpler API if we're keeping it as one strategy though - coercing bounds to UTC means we can't generate the values that usually cause overflows, dropping `allow_naive` means you have to compose strategies to get mixed values (also a rich source of bugs), and timezones has the same problem plus we'd need a default for aware bounds.

/s/timezone/datetime/

The other (ie not what I've done here) sensible API design would be to split `datetimes()` into two functions, `naive_datetimes()` and `aware_datetimes()`.  This would be a lot cleaner, but require anyone who wanted both to compose them explicitly.  Which, darn, is probably worth it - I'll copy this to a top-level comment for discussion.